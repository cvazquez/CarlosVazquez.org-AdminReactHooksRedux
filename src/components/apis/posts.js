import { checkAPIResponse } from '../../helpers/api'

// API request blog posts, to display and edit
export function getPosts() {
	return new Promise(resolve => {
		fetch(`${process.env.REACT_APP_API_URL}/getPosts`)
		.then(res => checkAPIResponse(res))
		.then(
			result => {
				if(result.posts && Array.isArray(result.posts)) {
					// Update render state to display active entries only and replace loading text
					resolve({
						isLoaded		: true,
						activeEntries	: result.posts.filter(entry => entry.deletedAt === null),
						isAdmin			: result.isAdmin
					});


				} else {
						throw new Error("Result posts response is invalid. Check API response")
				}

			},
			error => {
				resolve({
					isLoaded	: false,
					error
				});

				console.log("No Response from API to retrieve posts", error)
			}
		)
		.catch(error => {
			resolve({
				isLoaded	: false,
				error
			});

			console.error("API Request Fetch Error:", error)
		})
	})

}