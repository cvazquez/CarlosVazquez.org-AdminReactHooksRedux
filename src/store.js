import rootReducer from "./reducer";
import { configureStore } from "@reduxjs/toolkit";

const 	store = configureStore({
    reducer: rootReducer,
    // eslint-disable-next-line
    devTools: process.env.NODE_ENV !== 'production',
});

export default store;