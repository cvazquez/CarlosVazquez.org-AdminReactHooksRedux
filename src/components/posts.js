/*
	 Display a list of Blog posts to Edit/Deactivate

	 TODOS:
		 * Add ability to show deactivated posts
		 * Add ability to sort title, id and date columns
 */

import React, { useState, useEffect, useRef } from "react";
import { Link } from "react-router-dom"
import useDeactivateModal from "./modals/useDeactivate";
import { getPosts } from "./apis/posts";
import { showDemoMessage } from "../helpers/login";
import { useSelector } from "react-redux";


export default function Posts() {
    const	[apiResponses, setAPIResponses] = useState({
            // Async check if posts have loaded
            error		: null,
            isLoaded	: false,
            isAdmin		: true, // temporary (fake) security for demoing

            // Posts loaded from API call
            activeEntries	: []
        }),
        deactivateModal = useDeactivateModal(),
        unMounted		= useRef(false),
        posts 			= useSelector(state => state.posts);

    deactivateModal.referenceAPIResponses(setAPIResponses);

    useEffect(() => {
        async function getData(posts) {
            let data = await posts;

            if(posts.isLoaded) {
                data = posts;
                data.activeEntries = data.entries;
            } else {
                data = await getPosts();
            }

            data.isLoaded ?
                setAPIResponses(state => ({
                    ...state,
                    isLoaded		: data.isLoaded,
                    activeEntries	: data.activeEntries,
                    isAdmin			: data.isAdmin
                }))
                :
                setAPIResponses(state => ({
                    ...state,
                    isLoaded	: false,
                    error		: data.error
                }));
        }

        if(!unMounted.current) {
            getData(posts);
        }

        return () => unMounted.current = true;
    }, [posts])

    // When a post's Deactivate link is clicked
    function handleDeleteClick(e) {
        const	post	= e.currentTarget.dataset; // Title and id of the post clicked

        // Add an opaque overlay over entire body
        document.body.classList.add("overlay");

        deactivateModal.setModal(modal => ({
            ...modal,
            activate	: true,
            title		: post.title,
            id			: post.id,
            entries		: apiResponses.activeEntries
        }));
    }


    if (apiResponses.error) {

        // Display error message from the API call
        return <div>Error: {apiResponses.error.message}</div>;

    } else if (!apiResponses.isLoaded) {

        // While the API call is waiting for a response
        return <div>Loading...</div>;

    } else {
    // API call response is successful. Display list of posts.
    // display list of posts to edit/deactivate
        return <>
            {	// When Deactivate is clicked, display the confirmation modal
                deactivateModal.display()
            }
            {showDemoMessage(!apiResponses.isAdmin)}

            <div className="grid-container posts">
                <div className="grid-header-wrapper">
                    <div>Id</div>
                    <div>Title</div>
                    <div>Actions</div>
                    <div>Created</div>
                    <div>Published</div>
                    <div>Deleted</div>
                </div>
                {	// Loop through list of posts and display, with id, title, dates and links to Edit/Deactivate,
                    apiResponses.activeEntries.map(entry => (
                        <React.Fragment key={entry.id}>
                            <div className="grid-row-wrapper"
                            >
                                <div>{entry.id}</div>
                                <div>{entry.title}</div>
                                <div>
                                    <Link	to			= {`/posts/edit/${entry.id}`}
                                        key			= {`Entry${entry.id}`}
                                        className	= "edit"
                                        data-testid	= {entry.id}>
                                        Edit&nbsp;
                                    </Link>
                                    <span
                                        onClick		= {handleDeleteClick}
                                        onKeyDown   ={handleDeleteClick}
                                        data-id		= {entry.id}
                                        data-title	= {entry.title}
                                        className	= "delete"
                                        role        = "button"
                                        tabIndex    = {0}>
                                        Deactivate
                                    </span>
                                </div>
                                <div>{entry.createdAt}</div>
                                <div>{entry.publishAt}</div>
                                <div>{entry.deletedAt}</div>
                            </div>
                        </React.Fragment>
                    ))
                }
            </div>
        </>
    }
}