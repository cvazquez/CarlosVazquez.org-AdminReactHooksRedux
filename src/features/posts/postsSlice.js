//import axios from "axios";
import { getPosts } from "../../components/apis/posts";


// Current state of tabs editing
const initialState = {
    isLoaded	: false,
    entries		: [],
    isAdmin		: false
};

export default function postsReducer(state = initialState, action) {
    switch(action.type) {
        case "posts/get":

            return {
                ...state,
                isLoaded	: action.payload.isLoaded,
                entries		: action.payload.activeEntries,
                isAdmin		: action.payload.isAdmin
            }
        default:
            return state
    }
}

export async function fetchPosts(dispatch) {
    const data = await getPosts();


    //dispatch({ type: 'posts/get', payload : data.isLoaded ? data.activeEntries : [] })
    dispatch({ type: "posts/get", payload : data });
}
