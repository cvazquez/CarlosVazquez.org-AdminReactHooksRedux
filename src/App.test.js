import React from "react";
import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import App, {getLinks} from "./App";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";

import store from "./store";
import { Provider } from "react-redux";
import { fetchPosts } from "./features/posts/postsSlice";



afterEach(cleanup);

it("renders without crashing", () => {
    render(<App />, <div />);
});

it("should take a snapshot", () => {
    const { asFragment } = render(<App />);

    expect(asFragment(<App />)).toMatchSnapshot();
});

test("Has Home Text", () => {
    const { getByText } = render(<App />);
    const linkElement = getByText(/Home/);
    expect(linkElement).toBeInTheDocument();
});

test("Has Post Text", () => {
    const { getByText } = render(<App />);
    const linkElement = getByText(/Post/);
    expect(linkElement).toBeInTheDocument();
});

test("Links should match", () => {
    const obj = getLinks();

    expect(obj).toEqual(
        expect.objectContaining([{
            path	: "/",
            text	: "Home"
        },
        {
            path	: "/posts",
            text	: "Posts"
        },
        {
            path	: "/posts/add",
            text	: "Add"
        },
        {
            path	: "/categories",
            text	: "Categories"
        },
        {
            path	: "/series",
            text	: "Series"
        }])
    );
});


const renderWithRouter = (component) => {
    const history = createMemoryHistory();
    return {
        ...render (
            <Router history={history}>
                {component}
            </Router>
        )
    };
};

it("should render the home page", () => {

    const { container, getByTestId } = renderWithRouter(<App />);
    const navbar = getByTestId("header");
    const link = getByTestId("Posts");

    expect(container.innerHTML).toMatch("Home");
    expect(navbar).toContainElement(link);
});

// /posts
it('should navigate to the posts page loading', ()=> {
    const unsubscribe = store.subscribe(() =>
        // eslint-disable-next-line no-console
        console.log('State after dispatch: ', store.getState())
    );
    store.dispatch(fetchPosts);
    unsubscribe();


    const { container, getByTestId } = renderWithRouter(<Provider store={store}><App /></Provider>);

    fireEvent.click(getByTestId('Posts'));

    expect(container.innerHTML).toMatch('Loading...');
});

it('should navigate to the posts page loaded', async () => {
    const { getByText , getByTestId } = renderWithRouter(<App />);

    fireEvent.click(getByTestId('Posts'));

    const posts = await waitFor(() => getByText("Title"));

    expect(posts).toHaveTextContent('Title');
});

// /posts/add
it("should navigate to the add page loading", ()=> {
    const { container, getByTestId } = renderWithRouter(<App />);

    fireEvent.click(getByTestId("Add"));

    expect(container.innerHTML).toMatch("Loading...");
});

it("should navigate to the add page loaded", async () => {
    const { getByTestId } = renderWithRouter(<App />);

    fireEvent.click(getByTestId("Add"));

    const add = await waitFor(() => getByTestId("Save"));

    expect(add).toHaveAttribute("type", "submit");
}, 10000);

// /posts/edit
it('should navigate to the edit page loading', async ()=> {
    const { container, getByTestId, getByText } = renderWithRouter(<App />);

    fireEvent.click(getByTestId('Posts'));

    await waitFor(() => getByText("Title"));

    fireEvent.click(getByTestId('38'));

    expect(container.innerHTML).toMatch('Loading...');
});

it('should navigate to the edit page loaded', async ()=> {
    const { getByTestId, getByText } = renderWithRouter(<App />);

    fireEvent.click(getByTestId('Posts'));

    await waitFor(	() => getByText("Title"));

    fireEvent.click(getByTestId('38'));

    const edit = await waitFor(() => getByTestId("Save"));

    expect(edit).toHaveAttribute('type', 'submit');
}, 10000);

// /categories
it("should navigate to the categories page loading", ()=> {
    const { container, getByTestId } = renderWithRouter(<App />);

    fireEvent.click(getByTestId("Categories"));

    expect(container.innerHTML).toMatch("Loading...");
});

it("should navigate to the categories page loaded", async ()=> {
    const { getByTestId } = renderWithRouter(<App />);

    fireEvent.click(getByTestId("Categories"));

    const posts = await waitFor(() => getByTestId("About Me"));

    expect(posts).toHaveClass("Category");
});

// /series
it("should navigate to the series page loading", ()=> {
    const { container, getByTestId } = renderWithRouter(<App />);

    fireEvent.click(getByTestId("Series"));

    expect(container.innerHTML).toMatch("Loading...");
});

it("should navigate to the series page loaded", async ()=> {
    const { getByTestId } = renderWithRouter(<App />);

    fireEvent.click(getByTestId("Series"));

    const	series			= await waitFor(() => getByTestId("Patagonia")),
        seriesManage	= getByTestId("Patagonia_manage");

    expect(series).toHaveClass("Series");
    expect(seriesManage).toHaveClass("series-manage-click");
});