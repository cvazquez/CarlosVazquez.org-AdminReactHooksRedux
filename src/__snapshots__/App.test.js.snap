// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <ul
    class="header"
    data-testid="header"
  >
    <li>
      <a
        data-testid="Home"
        href="/"
      >
        Home
      </a>
    </li>
    <li>
      <a
        data-testid="Posts"
        href="/posts"
      >
        Posts
      </a>
    </li>
    <li>
      <a
        data-testid="Add"
        href="/posts/add"
      >
        Add
      </a>
    </li>
    <li>
      <a
        data-testid="Categories"
        href="/categories"
      >
        Categories
      </a>
    </li>
    <li>
      <a
        data-testid="Series"
        href="/series"
      >
        Series
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
